pkg/codegen/nodejs/gen.go

```diff
@@ -1617,6 +1684,223 @@ type namespace struct {
+		ns.name,
+		len(ns.types),
+		len(ns.enums),
+		childrenStr,
+	)
+}
+
+// ioContext defines a set of parameters used when generating input/output
+// type definitions. These parameters are stable no matter which directory
+// is getting generated.
+type ioContext struct {
+	mod             *modContext
+	input           bool
+	imports         map[string]codegen.StringSet
+	externalImports codegen.StringSet
+}
+
+// filename returns the unique name of the input/output file
+// given a directory root.
+func (ctx *ioContext) filename(dirRoot string) string {
+	var fname = fmt.Sprintf("%s.ts", ctx.filetype())
+	return path.Join(dirRoot, fname)
+}
+
+func (ctx *ioContext) filetype() string {
+	if ctx.input {
+		return "input"
+	}
+	return "output"
+}
+
+// intoIOFiles converts this namespace into one or more files.
+// It recursively builds one file for each node in the tree.
+// If input=true, then it builds input types. Otherwise, it
+// builds output types.
+// The parameters in ctx are stable regardless of the depth of recursion,
+// but parent is expected to change with each recursive call.
+func (ns *namespace) intoIOFiles(ctx *ioContext, parent string) ([]*ioFile, error) {
+	// We generate the input and output namespaces when there are enums,
+	// regardless of whether they are empty.
+	if ns == nil {
+		return nil, fmt.Errorf("Generating IO files for a nil namespace")
+	}
+
+	// We want to organize the items in the source file by alphabetical order.
+	// Before we go any further, sort the items so downstream calls don't have to.
+	ns.sortItems()
+
+	// Declare a new file to store the contents exposed at this directory level.
+	var dirRoot = path.Join(parent, ns.name)
+	var file, err = ns.genOwnedTypes(ctx, dirRoot)
+	if err != nil {
+		return nil, err
+	}
+	var files = []*ioFile{file}
+
+	// We have successfully written all types at this level to
+	// input.ts/output.ts.
+	// Next, we want to recurse to the next directory level.
+	// We also need to write the index.file at this level (only once),
+	// and when we do, we need to re-export the items subdirectories.
+	children, err := ns.genNestedTypes(ctx, dirRoot)
+	if err != nil {
+		return files, err
+	}
+	files = append(files, children...)
+	// Lastly, we write the index file for this directory once.
+	// We don't want to generate the file twice, when this function is called
+	// with input=true and again when input=false, so we only generate it
+	// when input=true.
+	// As a special case, we skip the top-level directory /types/, since that
+	// is written elsewhere.
+	if parent != "./types" && ctx.input {
+		var indexFile = ns.genIndexFile(ctx, dirRoot)
+		files = append(files, indexFile)
+	}
+	return files, nil
+}
+
+// sortItems will sort each of the internal slices of this object.
+func (ns *namespace) sortItems() {
+	sort.Slice(ns.types, func(i, j int) bool {
+		return tokenToName(ns.types[i].Token) < tokenToName(ns.types[j].Token)
```

Comment:
This sort is possibly non-deterministic again (see my earlier comment). What's the the of `ns.types`?

Reply:
Good catch! Thanks!